Add digital credentials

rhysyngsun opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
Closes #1951 
Closes #1969 

#### What's this PR do?
This PR encompasses _a lot_ of changes. If the reviewer would like I can probably split some of the changes out into a separate PR, it'll just take some more time to do so. Overall, this PR implements Digital Credentials for xPro:

- Adds `mitol-django-oauth-toolkit-extensions` - to support digital-credentials-specific oauth scopes
- Adds `mitol-django-digital-credentials`
  - Necessary configuration and related integrations added
- Added signal on `ProgramCertificate` and `CourseRunCertificate` creation to trigger a digital credential email
- Styles the `/oauth2/authorize` page using our standard look & feel
  - Refactored JS settings rendering out into a template tag
  - There was some additional cleanup needed to fix circular references, several functions were moved into `mitxpro/utils.py` as a result, and i added test coverage on those as they weren't covered directly
 
#### How should this be manually tested?

Enable the features `FEATURE_DIGITAL_CREDENTIALS` and `FEATURE_DIGITAL_CREDENTIALS_EMAIL` in `.env`

##### In django admin, create an oauth application with the following settings:

- Redirect URI: `http://localhost:5000/redirect_complete` (this is arbitrary)
- Client Type: `Public`
- Grant Type: `Authorization Code`
- Skip Authorization: `No`

##### Then, run this script inside a docker python shell:

```python
from urllib.parse import *

from django.contrib.contenttypes.models import ContentType
from courses.credentials import send_digital_credential_request_notification
from courses.factories import *
from courses.models import *
from mitol.digitalcredentials.models import *

user = User.objects.get(email="<EMAIL>)
program = Program.objects.last()
certificate, _ = ProgramCertificate.objects.get_or_create(program=program, user=user)
CourseRunCertificate.objects.get_or_create(course_run=program.courses.first().courseruns.first(), user=user)

credential_request, created = DigitalCredentialRequest.objects.get_or_create(
    credentialed_object_id=certificate.id,
    credentialed_content_type=ContentType.objects.get_for_model(certificate),
    learner=certificate.user,
)

if not created:
    send_digital_credential_request_notification(credential_request)
```
You should received an email notification that contains a deep link. This is arbitrary and likely to change very soon so it's not a big concern/priority.

Next, navigate to `http://xpro.odl.local:8053/oauth2/authorize/?client_id=<OAUTH_CLIENT_ID>&redirect_uri=http%3A%2F%2Flocalhost%3A5555%2Fdone&response_type=code`, this will redirect to the callback/redirect url, which you copy/paste into the next block of code.

##### Run the rest of these steps in a shell OUTSIDE docker
```python
import requests
from urllib.parse import *
#
client_id = "<OAUTH CLIENT ID>"
credential_request_url = "<CREDENTIAL REQ URL>"
token_url = "<TOKEN URL>"
challenge = "<CHALLENGE>"

# this is the only thing that changes between subsequent runs
callback_url= "<CALLBACK_URL>"

# NOTE: this is a one time code, if you consume it you'll need to start over at 1)
code = parse_qs(urlparse(callback_url).query)["code"]

# 3) redeem the authorization_code
data = dict(grant_type="authorization_code", code=code, client_id=client_id, scope="read write digitalcredentials")
token_resp = requests.post(token_url, data=data)  
access_token = token_resp.json()["access_token"]

# generate a proof of control of the DID
did = "did:example:456"
proof_resp = requests.post(
  "http://localhost:5000/generate/controlproof",
  json={
    "presentationId": did,
    "holder": "did:web:digitalcredentials.github.io",
    "verificationMethod": "did:web:digitalcredentials.github.io#96K4BSIWAkhcclKssb8yTWMQSz4QzPWBy-JsAFlwoIs",
    "challenge": challenge
  }
)

cred_resp = requests.post(
    credential_request_url,
    headers={
        "Authorization": f"Bearer {access_token}"
    },
    json=proof_resp.json()
)

print(cred_resp.json())
```

#### Screenshots (if appropriate)
(ignore the broken logo, that's a local development quirk)
![Screenshot_2021-01-25 MIT xPRO Digital Credentials Available - rhysyngsun gmail com - Gmail](https://user-images.githubusercontent.com/28598/105779456-6f9a6d00-5f3c-11eb-99e7-9a0946d6db12.png)

The user gets this confirmation page when attempting to authenticate the wallet app, I just used our standard auth ui
![Screenshot_2021-01-25 MIT xPRO](https://user-images.githubusercontent.com/28598/105779459-70330380-5f3c-11eb-8341-f20b5a0a3c06.png)

pdpinch commented on courses/credentials.py:
```diff
@@ -0,0 +1,92 @@
+        "identifier": program_run.program.text_id,
+        "name": program_run.program.title,
+        "description": program_run.program.page.description,
+        "numberOfCredits": {"value": program_run.program.page.certificate_page.CEUs},
```
Since we have some data coming from the CMS, can we raise an error if these happen to be empty?

pdpinch commented on courses/credentials.py:
```diff
@@ -0,0 +1,92 @@
+        "description": program_run.program.page.description,
+        "numberOfCredits": {"value": program_run.program.page.certificate_page.CEUs},
+        "startDate": program_run.start_date,
+        "endDate": program_run.end_date,
```
Probably should validate these as well, as I think they are optional fields.

pdpinch commented on courses/credentials.py:
```diff
@@ -0,0 +1,92 @@
+        "courseCode": course_run.course.readable_id,
+        "name": course_run.course.title,
+        "description": course_run.course.page.description,
+        "numberOfCredits": {"value": course_run.course.page.certificate_page.CEUs},
```
Same as for the program, I think we should raise an error if these fields are empty.

pdpinch reviewed (commented):
The schema and field mapping looks 👍 to me. 

I've got some concern about validation errors, since we're depending on CMS (and other) data that might be missing. I'm not sure the best time and place to handle this.

rhysyngsun commented:
@pdpinch I'm thinking we should error when the certificate is requested, but also preemptively error before we send emails to learners so we can fix obvious errors before learners try to get a certificate.

mbertrand reviewed (changes requested):
Some initial first pass comments on code.  I tried stepping through functional testing anyway and got this error when logging in via the authentication url generated in step 1: 'invalid_request, Mismatching redirect URI.'

rhysyngsun commented on courses/credentials.py:
```diff
@@ -0,0 +1,167 @@
+log = logging.getLogger(__name__)
+
+
+def build_program_credential(certificate: ProgramCertificate) -> Dict:
```
I don't think we've made a team decision on that, but i think it duplicate efforts typing-wise. I'm torn on whether there's benefit to describing the parameters cause speaking for myself I find the parameter descriptions are just sentence versions of the parameter name

rhysyngsun commented:
@mbertrand I've sorted most things here, what i've done since you last reviewed:

- Added the verifications on program/course/page configurations to be sure we don't issue certificates for a program that's not setup per Peter's comments earlier
- Updated the template to work against the newest `sign-and-verify` service (make sure you `git pull` + `npm run build` + `npm run start` to get the latest)
- Wrote a small PR https://github.com/mitodl/ol-django/pull/26 to adjust the view permissions so the view doesn't require write permissions - I'll need to release this and update the dependencies in this project before that can be used, but in the meantime you can just configure the application scopes to include the read/write perms too

I still need to update `.env.example` plus add some documentation for this, but i'll take care of those tomorrow

mbertrand commented on docs/configure_digital_credentials.md:
```diff
@@ -0,0 +1,24 @@
+| `DIGITAL_CREDENTIALS_ISSUER_ID` | - | The digital credentials issuer id to be included in the credential template. This value is determined by the digital credentials team. |
+| `DIGITAL_CREDENTIALS_VERIFICATION_METHOD` | - | The digital credentials verification method to be included in the credential template. This value is determined by the digital credentials team. |
+| `MITOL_DIGITAL_CREDENTIALS_HMAC_SECRET` | - | Any random string, MUST match the corresponding HMAC secret `sign-and-verify` is deployed with. |
+| `MITOL_DIGITAL_CREDENTIALS_VERIFY_SERVICE_BASE_URL` | `http://sign-and-verify.example.com:5678/` | The addressable base url for the `sign-and-verify` service. |
```
Unless there will be a ci/rc service we can use, maybe add documentation on how to get the local service running? `git clone digitalcredentials/sign-and-verify` , `npm run build & npm run start`, make accessible from inside docker, etc

mbertrand reviewed (changes requested):
Looking good!

rhysyngsun commented:
@mbertrand this should be all set for one last review

rhysyngsun commented:
Resolved conflicts.

mbertrand reviewed (commented):
I'm getting an error again but my mac is getting increasingly flaky so I'll 👍 if it's working for you, code looks fine

mbertrand commented:
This is the error I'm getting now, final step:

```
web_1     | [2021-02-09 20:43:33] ERROR 18 [django.request] log.py:228 - [2360535cca88] - Internal Server Error: /api/v1/credentials/request/e45fea23-1f9b-415a-b168-0ef923fb007a/
web_1     | Traceback (most recent call last):
web_1     |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
web_1     |     response = get_response(request)
web_1     |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 115, in _get_response
web_1     |     response = self.process_exception_by_middleware(e, request)
web_1     |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 113, in _get_response
web_1     |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
web_1     |   File "/usr/local/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
web_1     |     return view_func(*args, **kwargs)
web_1     |   File "/usr/local/lib/python3.7/site-packages/django/views/generic/base.py", line 71, in view
web_1     |     return self.dispatch(request, *args, **kwargs)
web_1     |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 495, in dispatch
web_1     |     response = self.handle_exception(exc)
web_1     |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 455, in handle_exception
web_1     |     self.raise_uncaught_exception(exc)
web_1     |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 492, in dispatch
web_1     |     response = handler(request, *args, **kwargs)
web_1     |   File "/usr/local/lib/python3.7/site-packages/mitol/digitalcredentials/views.py", line 42, in post
web_1     |     serializer.is_valid(raise_exception=True)
web_1     |   File "/usr/local/lib/python3.7/site-packages/rest_framework/serializers.py", line 236, in is_valid
web_1     |     self._validated_data = self.run_validation(self.initial_data)
web_1     |   File "/usr/local/lib/python3.7/site-packages/rest_framework/serializers.py", line 437, in run_validation
nginx_1   | 172.23.0.1 - - [09/Feb/2021:20:43:33 +0000] "POST /api/v1/credentials/request/e45fea23-1f9b-415a-b168-0ef923fb007a/ HTTP/1.1" 500 29409 "-" "python-requests/2.25.1" "-"
web_1     |     value = self.validate(value)
web_1     |   File "/usr/local/lib/python3.7/site-packages/mitol/digitalcredentials/serializers.py", line 47, in validate
web_1     |     log.debug("Failed to verify presentation: %s", result.json())
web_1     |   File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 897, in json
web_1     |     return complexjson.loads(self.text, **kwargs)
web_1     |   File "/usr/local/lib/python3.7/json/__init__.py", line 348, in loads
web_1     |     return _default_decoder.decode(s)
web_1     |   File "/usr/local/lib/python3.7/json/decoder.py", line 337, in decode
web_1     |     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
web_1     |   File "/usr/local/lib/python3.7/json/decoder.py", line 355, in raw_decode
web_1     |     raise JSONDecodeError("Expecting value", s, err.value) from None
```

mbertrand commented:
Gave it another try today, temporarily got around the migration issue by removing the outdated dependency from the migration script.  The link in the generated email has a null client_id (`http://xpro.odl.local:8053/oauth2/authorize/?client_id=None&response_type=code&redirect_uri=.......`)

rhysyngsun commented:
@mbertrand I've updated the implementation to largely revert back to a previous version of it per discussion with the digital credentials team. This now aligns to this document here: https://github.com/digitalcredentials/docs/blob/main/request/credential_request.md#1-initiate-request

Additionally, we can de-emphasize the email notification since there's some TBD UX changes around that which will probably result in a different delivery method for that link.

This also no longer is blocked by the oauth toolkit upgrade requirement. I also think that last issue you commented about won't be an issue because the code for that isn't present anymore.

mbertrand reviewed (approved):
👍
